### PR TITLE
Add CSP compatible parsing mode

### PR DESCRIPTION
**Pull request by MartinKolarik**
This adds a new parser option `insertExpressions` (name suggestions welcome). When set to true, generated functions are stored under `e` property of the template and no `eval` is needed on run time.

Refs: #1897, #2286.

### CODE REVIEW

File: src/parse/utils/insertExpressions.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	while ( i-- ) {
+		args[i] = `_${i}`;
+	}
+
+	// Functions created directly with new Function() look like this:
+	//     function anonymous (_0 /**/) { return _0*2 }
+	//
+	// With this workaround, we get a little more compact:
+	//     function (_0){return _0*2}
+	return new Function( [], `return function (${args.join(',')}){return(${x.s})}` )();
```
Comment (fskreuz) [inline src/parse/utils/insertExpressions.js]: Isn't the whole idea of CSP avoiding the use of `Function`? Or am I missing something here?

Comment (martypdx) [inline src/parse/utils/insertExpressions.js]: This module is called at parse time (see previous change), not during render.

File: src/shared/getFunction.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export let functions = {};
+
+export default function getFunction ( str, i ) {
+	if ( functions[ str ] ) return functions[ str ];
+
+	let args = new Array( i );
+	while ( i-- ) args[i] = `_${i}`;
+
+	return ( functions[ str ] = new Function( args.join( ',' ), `return (${str})` ) );
```
Comment (martypdx) [inline src/shared/getFunction.js]: But this isn't going to work because we're back to creating Functions.

**Comment by martypdx** [comment]
I think we need to store the function itself, not the string representation as `template.e`.

Then we should also be able to move the function creation logic solely into the parser. So everything still works for client side parsing, but if just using ractive runtime than there are no Function constructor calls.

At that point, wouldn't we be able to just make that the default behavior?

File: src/shared/getFunction.js (hunk 2 above)
Comment (fskreuz) [inline src/shared/getFunction.js]: Exactly my point. The functions are still strings.

**Comment by fskreuz** [comment]
So if I understand correctly now, to avoid `Function`, one must precompile the templates beforehand and the pre-compiled expressions are already functions, not just the `_0+_1` that need to be revived by `Function`. If so, sounds like a plan.

**Comment by evs-chris** [comment]
One of the things I was going to try out was a standalone precompiler that took a regular old template and added `fn: function(...) { ... }` to all of the expression nodes. Then a simple check for an `fn` key would let Ractive know whether or not to create a new function.

File: src/shared/getFunction.js (hunk 2 above)
Comment (martypdx) [inline src/shared/getFunction.js]: I think @MartinKolarik has it covered, we just need to (re)move this last line, or move it into the parser. Trying it out now...

**Comment by martypdx** [comment]
> One of the things I was going to try out was a standalone precompiler that took a regular old template and added fn: function(...) { ... } to all of the expression nodes.

I like @MartinKolarik's approach of coalescing all the functions to the template root and looking them up by key. But maybe `fn` would be better name for node?

I moved the `new Function` out of `getFunction` and instead added into to parser. Once I turned the functionality on globally, ran into some issues like `[[mustache]]` causing a function to be created (I beleive `r` uses `s` for static `true`).

File: src/shared/getFunction.js (hunk 2 above)
Comment (MartinKolarik) [inline src/shared/getFunction.js]: @martypdx no need to move anything, if functions are in the template, we don't get past 4th line. New function is created only if there isn't one (parsing without `insertExpressions`).

**Comment by MartinKolarik** [comment]
>I think we need to store the function itself, not the string representation as template.e.

That's what actually happens, though I admit the test may suggest otherwise - I was just trying to find a way to compare the functions for testing, since references will obviously differ.

>I beleive r uses s for static true

I added an `isArray()` check for `r`, and I believe this covers all cases. At first, I tried to use a more sophisticated function for finding expressions, but that got too complex too quickly, so I thought looking for the right keys anywhere in the templates might be a better idea. What do you think?

**Comment by JonDum** [comment]
> That's what actually happens, though I admit the test may suggest otherwise - I was just trying to find a way to compare the functions for testing, since references will obviously differ.

@MartinKolarik maybe this for the test?
```js
var a = function (_0,_1) { return _0 + _1 }
Object.toString.call(a)
< "function (_0,_1) { return _0 + _1 }"
```

**Comment by MartinKolarik** [comment]
@JonDum that's what I did initially, but it didn't work when running tests in phantom, so I changed that to just checking keys on `parsed.e`

**Comment by martypdx** [comment]
I moved the createFunction functionality completely into the parser and made pre-parsing functions the default behavior. All the tests still pass _except_ the expression parse tests which are testing the parsed result.

Before I fix those, I wanted to see if there was any reason _not_ to make this the default behavior. Also, I think we can write the functions as we go rather than post-processing them.

There's also some additional work to pre-write the computed property string.

**Comment by martypdx** [comment]
> I added an isArray() check for r, and I believe this covers all cases. At first, I tried to use a more sophisticated function for finding expressions, but that got too complex too quickly, so I thought looking for the right keys anywhere in the templates might be a better idea. What do you think?

@MartinKolarik working through some of the edge cases I changed the function to look explicitly for `.x` keys.

**Comment by MartinKolarik** [comment]
>Before I fix those, I wanted to see if there was any reason not to make this the default behavior. 

Increasing size of parsed templates for no real benefit (if you don't need this on).

>@MartinKolarik working through some of the edge cases I changed the function to look explicitly for .x keys.

It currently fails for `{{foo.bar[x + 6].baz}}` and `<a on-focus="method(x + 5)"></a>`

>There's also some additional work to pre-write the computed property string.

Computed properties are defined during run time, so there isn't much we can do about them, is there?

**Comment by JonDum** [comment]
> Before I fix those, I wanted to see if there was any reason not to make this the default behavior. Also, I think we can write the functions as we go rather than post-processing them.

@martypdx So the only downside is extra Kb in the template(s), but you save cpu cycles and battery and the JIT compiler *probably* has an easier time with these functions than with `new Function()`?

I'd have to do some testing to be sure, but I'd imagine it's such a small difference that it doesn't matter too much on cold load times. Especially if they're already bundled up in a request for the `js` instead of an additional request. 

Also, we'd avoid any future CSP issues that may arise in new version of browsers/iOS.

So :+1: for default from me.

**Comment by MartinKolarik** [comment]
I think the performance gains are just as negligible as the additional KBs, so I'm fine with whatever options you guys think is better. If we come to a conclusion that those extra KBs are insignificant compared to the rest of template, then we can even not make this an option, and simply always output all functions.

File: src/shared/getFunction.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import parser from '../Ractive/config/custom/template/parser';
+
+export let functions = {};
+
+export default function getFunction ( str, i ) {
+	if ( functions[ str ] ) return functions[ str ];
+
+	// Could we actually end up here???
+	return parser.createFunction( str, i );
```
Comment (MartinKolarik) [inline src/shared/getFunction.js]: If we leave this in, we'll need to cache those new functions.

**Comment by evs-chris** [comment]
I don't know if it's required anywhere, but I think there is something to be said for being able to have strict-JSON compatible templates.

**Comment by martypdx** [comment]
> I think the performance gains are just as negligible as the additional KBs

I think @MartinKolarik is probably right on the negligible nature of this. I went back to @Rich-Harris' original [perf investigation](https://gist.github.com/Rich-Harris/f3d63e4ef4d6680d3052) and while the mystery of the slighty faster _running_ generated function remains (at least on chrome), I measured the actual time to _create_ the generate function and it seems to run around .075 - .1 ms. So if you had 100 expressions, you might save 10ms on render. :D

I imagine anyone concerned about file size would be zipping, so I suspect the file sizes would be very similar to non-functions.

> I don't know if it's required anywhere, but I think there is something to be said for being able to have strict-JSON compatible templates.

Just to recap, without the precompiled functions, as long as the parser is available the expressions can be parsed on the client. So maybe we add a parser option of `csp: [true|false]` to control generating the functions. If set to false, it would require the parser version of Ractive on the client.

> It currently fails for {{foo.bar[x + 6].baz}} and <a on-focus="method(x + 5)"></a>

Hey, I borked that :)

> > There's also some additional work to pre-write the computed property string.
> Computed properties are defined during run time, so there isn't much we can do about them, is there?

Yeah, not entirely sure what best course is. One course is to require parser if you're going to use them. Though here is that parser === non-csp. You could add posted-processing of component build process that translates them. Though we'd have to expose that through parser somehow.

**Comment by MartinKolarik** [comment]
>Yeah, not entirely sure what best course is. One course is to require parser if you're going to use them. Though here is that parser === non-csp. You could add posted-processing of component build process that translates them. Though we'd have to expose that through parser somehow.

If you use CSP, you can just use real functions instead of strings. If you don't use CSP, nothing changes, and those functions will be created on run time. They are created by different function, so no parser is necessary.

Only possible problem I see is that someone might get confused, because setting `csp: true` won't (can't) do anything about those functions.

Handling this during build process *is* a way, but... is it worth it for saving those few characters? I personally declare all computed properties as real functions.

**Comment by MartinKolarik** [comment]
@martypdx are we good to merge this?

**Comment by martypdx** [comment]
@MartinKolarik I've got a few changes to push that allow the computation signatures to work without the parser, but after thinking about it I think that deploying the ractive runtime should _always_ be csp compliant. Given that security is part technical and part perception, I think it would be better to know that using the non-parser runtime means having no `new Function` or `eval`.

I've added a method `Ractive.parse.computedStrings` that would make it a one-liner to update tooling like `rcu` to be able to convert computation strings:

```js
parse.computedStrings = function( computed ) {
	if ( !computed ) return [];

	Object.keys( computed ).forEach( key => {
		const value = computed[ key ];
		if ( typeof value === 'string' ) {
			computed[ key ] = getFunctionFromString( value );
		}
	});
};
```

So upside is no `csp` option to manage and including only the runtime is "guaranteed" no eval. (Both these things make the code simpler too).
 
The downside of this being that using the runtime-only with computation strings flat out won't work, but I bet 99.9% percent of people in the situation are using tooling to package components - and the burden will be to add the conversion line to the tooling.

Thoughts?

(@evs-chris: `JSON.stringify` omits function properties so AFAICT they would just be removed at that point. They templates would be functional assuming the parser was included with the runtime.)

**Pull request merged**